### ISSUE TITLE
CoPilot feature is really slow

### ISSUE THREAD

**Issue by tituslhy** [Chainlit/chainlit#1327]
**Describe the bug**
I'm not sure why but Chainlit CoPilot takes really long to render on the main app. I'm using a simple "Hello world" app and it still takes so long to render in my frontend.

**To Reproduce**

Chainlit app code:
```
import chainlit as cl

@cl.on_chat_start
async def on_chat_start():
    await cl.Message("Hello!").send()
    
@cl.on_message
async def on_message(message: cl.Message):
    response = await cl.Message(f"You said '{message.content}'").send()
    if cl.context.session.client_type == "copilot":
            fn = cl.CopilotFunction(
                name = "test",
                args={"message": response.content}
            )
            await fn.acall()
```

Command line code:
```
chainlit run app.py --port 80 --host 0.0.0.0 --headless --watch
```

nginx code:
```
worker_processes 2;
pid /var/run/nginx.pid;

error_log /var/log/nginx.error_log info;

events {
    worker_connections 1024;
    # use kqueue;
}

http{
    # This block configures a reverse proxy for a backend application named "ws-backend".

    upstream ws-backend {
      # Replace this with the actual IP address and port of your backend application.
      server localhost:80;
    }

    server {
      # This block defines a virtual server configuration in Nginx.
      # A virtual server can handle requests for a specific domain name
      # or IP address.

      # Listen on port 443 with SSL enabled. This means Nginx will
      # expect incoming connections on port 443 (standard HTTPS port)
      # and use Secure Sockets Layer (SSL) for secure communication.
      listen 443 ssl;

      # Replace these paths with the actual location of your SSL certificate
      # and key files. The certificate file contains the public key that
      # browsers use to encrypt communication, and the key file contains
      # the private key that your server uses to decrypt messages.
      # You can obtain an SSL certificate from a certificate authority (CA).
      ssl_certificate ...;
      ssl_certificate_key ...;

      # (Optional) You can uncomment this line to enforce HTTPS for all traffic.
      # When uncommented, Nginx will redirect any HTTP requests (port 80) to the HTTPS
      # version (port 443) using a 301 (Moved Permanently) status code.
      # return 301 https://$host$request_uri;

      # This section defines a location block that specifies how Nginx should handle
      # requests for a specific path within the server. Here, the location block applies
      # to all paths (`/`).
      location / {
        # This line tells Nginx to proxy all requests to the specified backend server defined in the 'ws-backend' upstream block.
        proxy_pass http://ws-backend/;

        # This line ensures the `Host` header in the forwarded request reflects the
        # original domain name accessed by the user. This is important for applications
        # that rely on the hostname.
        proxy_set_header Host $host;

        # This line sets the `X-Real-IP` header in the forwarded request to the client's
        # real IP address. This can be useful for applications that need to track the
        # origin of requests behind a proxy server.
        proxy_set_header X-Real-IP $remote_addr;

        # This line specifies the HTTP version to be used in proxied requests. 
        # Set to 1.1 for modern communication protocols. 
        proxy_http_version 1.1;

        # This line disables proxy caching for requests. This might be necessary
        # for applications that require dynamic content or frequent updates.
        proxy_cache_bypass $http_upgrade;

        # This line disables buffering of proxied responses. This can improve
        # performance for applications that send large amounts of data.
        proxy_buffering off;
      }
    }
}
```
```
sudo systemctl restart nginx
```
**Expected behavior**
I expected my app to load smoothly onto my frontend. But it's stuck with a lot of 'pending' web socket tasks. Once one task is finished it starts another web socket task. Even when the "web socket" task finishes, it still doesn't load smoothly (it freezes and I'm unable to type anything in the chat). 

I think that the key to this is to wait because when waiting it works. but it shouldn't be like that - when users open the copilot they want to chat immediately.

**Screenshots**
Screenshot of my CoPilot app hanging
![image](https://github.com/user-attachments/assets/8378ef83-26b9-4c6a-8a1e-9a94450f1a61)

Screenshot of my "Network"
![image](https://github.com/user-attachments/assets/91855a92-fdca-4050-be06-8df26999310f)

Screenshot of my "Console"
![image](https://github.com/user-attachments/assets/eac8af43-153c-4c64-b5c7-5cbf44856710)

It's just a "hello world" app. It shouldn't fail.

**Comment by stephenrs** [comment]
@tituslhy I haven't looked at this in detail, but it looks like you have both nginx and chainlit listening on port 80, which will not work as expected.

So, it will probably work better to let chainlit run on the default port 8000 and allow nginx to reverse proxy requests to chainlit. I'm doing this in my project and you can see my working nginx configuration here #1262 

Also, depending on your goals, you might not need nginx to explicitly handle websocket connections because the chainlit server will handle them for you, although it's not clear how scalable it is.

**Comment by tituslhy** [comment]
Hmm but this shouldn't be the case right? Chainlit is not optimized for a specific port number (and another app is using that port haha). My app runs fine when it's a standalone app but on CoPilot it freezes. @willydouhard suspects that it might be because of the "--watch" command. I'll have to experiment to keep everyone updated.

**Comment by tituslhy** [comment]
I realize that it's actually very fast (or at least this "hello world" app) once I don't use nginx. This means I point my index.html to the un-routed URL link (http:<server>:<port>) instead of the nginx routed link sitting on port 443 (https://<server>). I'm really not sure why.

Also, when I deployed the index.html on localhost pointing to the nginx chainlit app it works ok too. It's only when I take the index.html off localhost and on an actual deployment that the freezing happens. So queer. I really don't understand why.

**Comment by willydouhard** [comment]
Is this still the case? We improved the copilot since then

**Comment by tituslhy** [comment]
I've not had the chance to experiment on it sorry! I'll close the issue first and raise another issue if I encounter this.
